include/array/z_order.h

```diff
@@ -0,0 +1,108 @@
+// Copyright 2019 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+/** \file z_order.h
+ * \brief Helpers for traversing multi-dimensional ranges in z-order.
+ */
+#ifndef NDARRAY_Z_ORDER_H
+#define NDARRAY_Z_ORDER_H
+
+#include "array.h"
+
+#include <algorithm>
+#include <functional>
+
+namespace nda {
+
+namespace internal {
+
+inline index_t next_power_of_two(index_t x) { 
+  index_t result = 1;
+  while (result < x) {
+    result <<= 1;
+  }
+  return result;
+}
+
+// TODO: This probably doesn't need fn to be inlined, and it would reduce template instantiations
+// to use std::function instead.
+template <size_t Rank, class Fn>
+NDARRAY_UNIQUE void for_each_index_in_z_order_impl(const std::array<index_t, Rank>& end,
+    std::array<index_t, Rank> z, index_t dim, index_t step, const Fn& fn) {
+  if (dim == 0 && step == 1) {
+    // We're at the innermost step of the traversal, call fn for the two neighboring points we have.
+    fn(z);
+    z[0] += 1;
+    if (z[0] < end[0]) fn(z);
```

Comment:
Nit, just for consistency.

```suggestion
    if (z[0] < end[0]) {
      fn(z);
    }
```